test/object-store/sync/sync_test_utils.hpp

```diff
@@ -50,6 +50,48 @@ void timed_wait_for(util::FunctionRef<bool()> condition,
 void timed_sleeping_wait_for(util::FunctionRef<bool()> condition,
                              std::chrono::milliseconds max_ms = std::chrono::seconds(30));
 
+template <typename T>
+struct TimedFutureState : public util::AtomicRefCountBase {
+    TimedFutureState(util::Promise<T>&& promise)
+        : promise(std::move(promise))
+    {
+    }
+
+    util::Promise<T> promise;
+    std::mutex mutex;
+    std::condition_variable cv;
+    bool finished = false;
+};
+
+template <typename T>
+util::Future<T> wait_for_future(util::Future<T>&& input, std::chrono::milliseconds max_ms)
```

Comment:
This works like timed_sleep_waiting_for that we use in a bunch of places, except it synchronously waits on a Future. If we like the way this works, I'll go through and clean up other places where we futz with std::atomic<bool>'s and the event loop to busy wait for a condition to become true.